src/stream/src/executor/upstream_sink_union.rs

```diff
@@ -0,0 +1,464 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use futures::future::try_join_all;
+use pin_project::pin_project;
+use risingwave_expr::expr::NonStrictExpression;
+
+use crate::executor::exchange::input::{Input, new_input};
+use crate::executor::prelude::*;
+use crate::executor::project::apply_project_exprs;
+use crate::executor::{BarrierMutationType, BoxedMessageInput, DynamicReceivers, MergeExecutor};
+use crate::task::{FragmentId, LocalBarrierManager};
+
+type MergeStream = impl Stream<Item = MessageStreamItem>;
+type ProcessedMessageStream = impl Stream<Item = MessageStreamItem>;
+
+#[pin_project]
+pub struct SinkHandlerInput {
+    /// The ID of the upstream fragment that this input is associated with.
+    upstream_fragment_id: FragmentId,
+
+    /// The stream of messages from the upstream fragment.
+    #[pin]
+    processed_stream: ProcessedMessageStream,
+}
+
+impl SinkHandlerInput {
+    pub fn new(
+        upstream_fragment_id: FragmentId,
+        merge: Box<MergeExecutor>,
+        project_exprs: Vec<NonStrictExpression>,
+    ) -> Self {
+        let merge_stream = Self::generate_stream_from_merge(merge);
+        let processed_stream = Self::apply_project_exprs_stream(merge_stream, project_exprs);
+        Self {
+            upstream_fragment_id,
+            processed_stream,
+        }
+    }
+
+    #[define_opaque(MergeStream)]
```

Comment:
Seems unnecessary to define the `MergeStream` opaque type. The `apply_project_exprs_stream` is called immediately after `generate_stream_from_merge`, so we can replace the parameter `merge_stream: MergeStream` with `merge: Box<MergeExecutor>`, and do `merge.execute_inner()` in `apply_project_exprs_stream_impl`.